### DISCUSSION TITLE
Auto rewrite/transform TypeScript import path from tsconfig.paths

### DISCUSSION THREAD

**Discussion by muuvmuuv** [biomejs/biome#2764] [Ideas]
Would be nice if Biome could auto fulfill import paths to tsconfig.compilerOptions.path. Maybe with a threshold option when it should do it:

```ts
import { IconComponent } from '../../component/icon'
```

becomes:

```ts
import { IconComponent } from '@/component/icon'
```

but:

```ts
import { IconComponent } from './icon'
import { IconComponent } from '../icon'
```

with a threshold of 1 wouldn't transform.

Since this would require a lot of I/O work it should be a flag/option.

**Comment by damianobarbati** [comment]
Does this exist now @muuvmuuv?

**Comment by ematipico** [comment]
I am not sure what this proposal is about. The title talks about formatting, but then the body explains that an import should be **transformed** into another path based on an external file.

Which one is it? The transformation doesn't seem feasible though, because it creates code that isn't portable.

**Reply by damianobarbati** [reply]
@ematipico here @muuvmuuv is referring to rewriting the paths respecting the `tsconfig.json`. I'm having the same problem lately. 

For instance, if I have my `Service.ts`
```ts
import Foe from "../../types/src/Foe.ts"
```

Then I change `tsconfig.json` to add:
```json
    "paths": {
      "@api/*": ["./services/api/src/*"],
      "@webapp/*": ["./services/webapp/src/*"],
      "@type/*": ["./services/types/src/*"]
    }
```

I would like `biome` to rewrite in `Service.ts`:
```ts
import Foe from "@types/Foe.ts"
```

Desiderata would be:
1. Biome rewriting paths respecting the subpath imports in package.json if any.
2. Biome rewriting paths respecting the tsconfig path in tsconfig.json if any.
3. @muuvmuuv suggests to add a threshold preventing the rewrite based on hierarchy: for instance `1` means don't rewrite if file referenced is in same directory, `2` if file referenced is in parent directory, etc... (the threshold is the `....` dots in hierarchy).

The point 1 is needed for the new `--experimental-strip-types flag` to work with relative paths: it won't support rewriting code as tsx does, so developers will probably (my guess) start using more and more the native subpath import feature of node.

**Reply by ematipico** [reply]
I see, thank you for detailed explanation.

We're still far away from being able to provide such functionality, as we miss the required infrastructure, so you'll have to wait a bit more.

**Reply by damianobarbati** [reply]
@ematipico definitely not urgent but a nice to have, thank you for the feedback! 🚀

**Reply by muuvmuuv** [reply]
Changed the title, format wasnt the correct naming :)

**Reply by toby-j** [reply]
Agreed, this would be a great feature

**Comment by damianobarbati** [comment]
I want to add another option here: I happily migrated to native nodejs type stripping several months ago, and typescript relative paths are an old memory.
I suggest to work on rewriting the paths respecting either:
- `package.json`: for those using native typescript and package.json relative imports in monorepo projects.
- - `tsconfig.json`: for those still using tsx and friends, and if relative paths are there.

To give you an idea => https://github.com/damianobarbati/ts-repro/blob/main/services/api/src/index.ts#L1

**Reply by muuvmuuv** [reply]
Didn't know about package.json imports :o will have a look at that, totally agree here

**Comment by siketyan** [comment]
I opened a task for this: https://github.com/biomejs/biome/issues/5936
